mjpc/planners/mppi/planner.cc

```diff
@@ -0,0 +1,508 @@
+                   MaxNoiseStdDev);
+
+  // add sampling planner
+  mjui_add(&ui, defSampling);
+}
+
+// planner-specific plots
+void MPPIPlanner::Plots(mjvFigure* fig_planner, mjvFigure* fig_timer,
+                        int planner_shift, int timer_shift, int planning,
+                        int* shift) {
+  // ----- planner ----- //
+  double planner_bounds[2] = {-6.0, 6.0};
+
+  // improvement
+  mjpc::PlotUpdateData(fig_planner, planner_bounds,
+                       fig_planner->linedata[0 + planner_shift][0] + 1,
+                       mju_log10(mju_max(improvement, 1.0e-6)), 100,
+                       0 + planner_shift, 0, 1, -100);
+
+  // legend
+  mju::strcpy_arr(fig_planner->linename[0 + planner_shift], "Improvement");
+
+  fig_planner->range[1][0] = planner_bounds[0];
+  fig_planner->range[1][1] = planner_bounds[1];
+
+  // bounds
+  double timer_bounds[2] = {0.0, 1.0};
+
+  // ----- timer ----- //
+
+  PlotUpdateData(
+      fig_timer, timer_bounds, fig_timer->linedata[0 + timer_shift][0] + 1,
+      1.0e-3 * noise_compute_time * planning, 100, 0 + timer_shift, 0, 1, -100);
+
+  PlotUpdateData(fig_timer, timer_bounds,
+                 fig_timer->linedata[1 + timer_shift][0] + 1,
+                 1.0e-3 * rollouts_compute_time * planning, 100,
+                 1 + timer_shift, 0, 1, -100);
+
+  PlotUpdateData(fig_timer, timer_bounds,
+                 fig_timer->linedata[2 + timer_shift][0] + 1,
+                 1.0e-3 * policy_update_compute_time * planning, 100,
+                 2 + timer_shift, 0, 1, -100);
+
+  // legend
+  mju::strcpy_arr(fig_timer->linename[0 + timer_shift], "Noise");
+  mju::strcpy_arr(fig_timer->linename[1 + timer_shift], "Rollout");
+  mju::strcpy_arr(fig_timer->linename[2 + timer_shift], "Policy Update");
+
+  // planner shift
+  shift[0] += 1;
+
+  // timer shift
+  shift[1] += 3;
+}
+
+double MPPIPlanner::CandidateScore(int candidate) const {
+  return trajectory[candidate].total_return;
+}
+
+// set action from candidate policy
+void MPPIPlanner::ActionFromCandidatePolicy(double* action, int candidate,
+                                            const double* state, double time) {
+  candidate_policy[candidate].Action(action, state, time);
+}
+
+void MPPIPlanner::CopyCandidateToPolicy(int candidate) {
+  {
+    const std::shared_lock<std::shared_mutex> lock(mtx_);
+    previous_policy = policy;
+
+    // the policy's action at time t is the previous policy's action at time t+1
+    // the last action is repeated
+    for (int t = 0; t < policy.num_spline_points - 1; t++) {
+      mju_copy(policy.parameters.data() + t * model->nu,
+               previous_policy.parameters.data() + (t + 1) * model->nu,
+               model->nu);
+    }
+    mju_copy(
+        policy.parameters.data() + (policy.num_spline_points - 1) * model->nu,
+        previous_policy.parameters.data() +
+            (policy.num_spline_points - 1) * model->nu,
+        model->nu);
+  }
```

Comment:
rolling the policy - there's definitely potential for this to be broken

Reply:
Yeah that seems quite possible. I don't grok the whole trajectory vs policy vs candidate policy vs previous policy thing. Just doing vanilla sampling with this setup could help isolate if this is the issue